### PR TITLE
Golang added mocks

### PR DESCRIPTION
**Pull request by michalzurawski**
This PR depends on #532 - please review only the last commit  (https://github.com/cloudwan/gohan/pull/535/commits/9e8bc0d84144833a34be5028a963a368224ea84f)

### CODE REVIEW

File: extension/framework/goplugin/runner/runner.go [hunk 1]
```diff
@@ -224,7 +225,9 @@ func (testRunner *TestRunner) runSingle(t ginkgo.GinkgoTestingT, reporter *Repor
 	}
 
 	// prepare test
-	test(env)
+	mockEnv.Reset()
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/framework/goplugin/runner/runner.go]: I think this is unceccesary. One reset after each test is enough. At the beginning, env should be clean.

File: extension/goext/environment_mock.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package goext
+
+//MockModules ...
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/goext/environment_mock.go]: Please fix comment

File: extension/goplugin/environment.go [hunk 3]
```diff
@@ -90,6 +90,23 @@ type Environment struct {
...
+	goext.IEnvironment
+	extension.Environment
+	getSchemaHandlers(event string) (SchemaPrioritizedSchemaHandlers, bool)
+	getHandlers(event string) (PrioritizedHandlers, bool)
+	dispatchSchemaEvent(prioritizedSchemaHandlers PrioritizedSchemaHandlers, sch Schema, event string, context map[string]interface{}) error
+	resourceFromContext(sch Schema, context map[string]interface{}) (res goext.Resource, err error)
+	getRawType(schemaID string) (reflect.Type, bool)
+	getType(schemaID string) (reflect.Type, bool)
+	updateResourceFromContext(resource interface{}, context goext.Context) error
+	RegisterSchemaEventHandler(schemaID string, event string, handler func(context goext.Context, resource goext.Resource, environment goext.IEnvironment) error, priority int)
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/goplugin/environment.go]: Could you sort public / private members and add one empty line to separate those sections?

File: extension/goplugin/environment_mock.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
+package goplugin
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/goplugin/environment_mock.go]: Add header

File: extension/goplugin/util.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package goplugin
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/goplugin/util.go]: Add header

File: extension/goplugin/util.go [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+import "github.com/golang/mock/gomock"
+
+var controllers map[gomock.TestReporter]*gomock.Controller
+
+func NewController(testReporter gomock.TestReporter) *gomock.Controller {
+	ctrl := gomock.NewController(testReporter)
+	if controllers == nil {
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/goplugin/util.go]: this initialization can be moved to declaration so the code here will be cleaner

**Comment by przemyslaw-dobrowolski-cl** [comment]
@michalzurawski I like the general idea - it's clean and does the job. Please fix some extra review  comments.

File: extension/framework/goplugin/runner/runner.go (hunk 1 above)
Comment (michalzurawski) [reply extension/framework/goplugin/runner/runner.go]: Currently this is required - without it every first test in the suite will fail. Added todo comment to remove it in the future.

**Comment by michalzurawski** [comment]
@przemyslaw-dobrowolski-cl  changes have been applied

File: extension/goplugin/environment_mock.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+		mockEnv.http = mockEnv.env.HTTP()
+	}
+
+	if mockEnv.mockModules.Auth {
+		mockEnv.auth = goext.NewMockIAuth(ctrl)
+	} else {
+		mockEnv.auth = mockEnv.env.Auth()
+	}
+
+	if mockEnv.mockModules.Util {
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/goplugin/environment_mock.go]: Util section should be last, not Config

File: extension/goplugin/generate_mocks.sh [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+file="environment_mock_gen.go"
+mockgen -package goext github.com/cloudwan/gohan/extension/goext ICore,ILogger,ISchemas,ISync,IDatabase,ITransaction,IHTTP,IAuth,IUtils,IConfig > ${file}
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/goplugin/generate_mocks.sh]: there is no IUtils

**Pull request merged**